src/rgw/rgw_rest.cc

```diff
@@ -382,11 +382,22 @@ void dump_header(struct req_state* const s,
   }
 }
 
+static inline boost::string_ref get_sanitized_hdrval(ceph::buffer::list& raw)
+{
+  /* std::string and thus boost::string_ref ARE OBLIGED to carry multiple
+   * 0x00 and count them to the length of a string. We need to take that
+   * into consideration and sanitize the size of a ceph::buffer::list used
+   * to store metadata values (x-amz-meta-*, X-Container-Meta-*, etags).
+   * Otherwise we might send 0x00 to clients. */
+  const size_t len = raw.length();
+  return boost::string_ref(raw.c_str(), len ? len - 1 : 0);
```

Comment:
okay, i see that we're converting strings to bufferlist like this: `bl.append(etag.c_str(), etag.size() + 1);` i'm not sure why we need a null terminator in these bufferlists, because we shouldn't be trying to print them out

can you make this trimming logic conditional, so that we only trim if they're `\0`s? that way we have the option to start removing those trailing nulls